### PR TITLE
[SPARK-58965][SQL][PYTHON] Add gcd and lcm math functions

### PR DESCRIPTION
**Pull request by Vivek1106-04**
### What changes were proposed in this pull request?

This PR adds two new built-in math functions, `gcd` and `lcm`:

```sql
gcd(expr1, expr2) -- returns BIGINT
lcm(expr1, expr2) -- returns BIGINT
```

Both arguments are implicitly cast to BIGINT and the result is BIGINT, matching the existing
`factorial` expression, which likewise takes an integral argument and returns BIGINT.

Semantics:

| Case | `gcd` | `lcm` |
|---|---|---|
| either argument is NULL | NULL | NULL |
| both arguments are 0 | 0 | 0 |
| one argument is 0 | `abs(other)` | 0 |
| negative arguments | result is non-negative | result is non-negative |
| result not representable as BIGINT | `ARITHMETIC_OVERFLOW` under ANSI mode, NULL otherwise | `ARITHMETIC_OVERFLOW` under ANSI mode, NULL otherwise |

The greatest common divisor is computed with the Euclidean algorithm. The least common multiple
divides by the greatest common divisor before multiplying, so a representable result never
overflows on the way there — for example `lcm(4611686018427387904, 2)` returns `4611686018427387904`
even though the naive product `4611686018427387904 * 2` would not fit.

Overflow arises in exactly two places, and in both PostgreSQL raises as well:

* `gcd` where the result would be `-Long.MinValue`, which is not representable — that is, the
  input pairs `(0, x)`, `(x, 0)` and `(x, x)` for `x = Long.MinValue`.
* `lcm` where `abs(a) / gcd(a, b) * abs(b)` exceeds `Long.MaxValue`.

Both are reported with the existing `ARITHMETIC_OVERFLOW` error condition, so no new error
condition is introduced. Overflow is gated on ANSI mode as it is elsewhere in Spark — `conv` in the
same file takes the same approach — raising under ANSI mode and returning NULL otherwise.

Both helpers live in `MathUtils` and are shared by the interpreted and codegen paths so the two
cannot diverge.

The functions are exposed through SQL, the Scala/Java `functions` API, PySpark, and Spark Connect.

### Why are the changes needed?

Spark SQL currently has no way to compute either value. There is no expression for it, and no
combination of existing built-ins produces the result, so users have to fall back to a UDF — which
for PySpark means a Python round trip per row and no whole-stage codegen.

Both functions are standard in comparable engines:

* PostgreSQL 13+: `gcd(a, b)`, `lcm(a, b)` for `integer`, `bigint` and `numeric`
* DuckDB: `gcd(a, b)`, `lcm(a, b)` (aliases `greatest_common_divisor`, `least_common_multiple`)

Common uses include reducing fractions and ratios to lowest terms, aligning batch or partition
sizes, computing the repeat period of overlapping schedules, and normalizing denominators before
aggregation.

### Does this PR introduce _any_ user-facing change?

Yes. Two new built-in functions are available in SQL, the Scala/Java `functions` API, PySpark and
Spark Connect. Previously `gcd` and `lcm` were unresolved function names:

```sql
-- Before
spark-sql> SELECT gcd(24, 36);
[UNRESOLVED_ROUTINE] Cannot resolve routine `gcd` on search path
[`system`.`builtin`, `system`.`session`, `spark_catalog`.`default`]. SQLSTATE: 42883

-- After
spark-sql> SELECT gcd(24, 36);
12
spark-sql> SELECT lcm(4, 6);
12
```

No existing behavior changes; the change is purely additive.

### How was this patch tested?

New tests:

* `MathExpressionsSuite` — unit tests for both expressions covering ordinary values, negative
  inputs, zero, NULL, the boundary values around `Long.MinValue` / `Long.MaxValue`, and overflow in
  both ANSI and non-ANSI mode, plus interpreted-vs-codegen consistency checks.
* `math.sql` golden-file tests, regenerated for the ANSI and non-ANSI results and the analyzer
  results.
* `PlanGenerationTestSuite` function tests, with the regenerated Spark Connect plan and explain
  golden files.
* PySpark doctests for `gcd` and `lcm`.

Existing suites re-run and passing: `MathExpressionsSuite`, `SQLQueryTestSuite` (`math.sql`),
`ExpressionsSchemaSuite`, `PlanGenerationTestSuite`, `ProtoToParsedPlanTestSuite`.

`dev/lint-scala` and `dev/lint-python` both pass.

### Was this patch authored or co-authored using generative AI tooling?

Generated-by: Claude Code (Claude Opus 5)

### CODE REVIEW

**Comment by Vivek1106-04** [comment]
@uros-b ! can you review this, added two new functions  which are standard functions in other databases.

**Comment by uros-b** [comment]
Thank you for @Vivek1106-04! I'll take a look soon. Also, cc @bojana-db to help with reviewing this PR.

File: sql/api/src/main/scala/org/apache/spark/sql/functions.scala [hunk 1]
```diff
@@ -5442,6 +5442,36 @@ object functions {
    */
   def factorial(e: Column): Column = Column.fn("factorial", e)
 
+  /**
```
Comment (uros-b) [inline sql/api/src/main/scala/org/apache/spark/sql/functions.scala]: `gcd`/`lcm` sit between `factorial` (5443) and `floor` (5488). The section is strictly alphabetical (cos, cosh, cot, csc, exp, expm1, factorial, floor, greatest, hex, hypot, least, ln, log), so `gcd` belongs after the `floor` overloads / before `greatest` (~5526) and `lcm` after the last `hypot` overload / before `least` (~5696).

`__init__.py` and `functions.rst` place both correctly, so this file is the lone inconsistency.

**Comment by Vivek1106-04** [comment]
I also double-checked the other places the functions are registered/exported: `FunctionRegistry.scala` is already alphabetical there (factorial, gcd, hypot, lcm), and the Python `functions/builtin.py` files aren't alphabetically ordered to begin with, so `functions.scala` was indeed the only inconsistency.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala [hunk 2]
```diff
@@ -762,6 +762,142 @@ case class Factorial(child: Expression)
...
+  override def nullIntolerant: Boolean = true
+
+  def this(left: Expression, right: Expression) =
+    this(left, right, ansiEnabled = SQLConf.get.ansiEnabled)
+
+  override def inputTypes: Seq[DataType] = Seq(LongType, LongType)
+
+  override def dataType: DataType = LongType
+
+  // The result overflows for inputs whose divisor is -Long.MinValue, which is null in ANSI mode.
```
Comment (bojana-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala]:
```suggestion
  // The result overflows for inputs whose divisor is -Long.MinValue, which is null in non-ANSI mode.
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala [hunk 3]
```diff
@@ -762,6 +762,142 @@ case class Factorial(child: Expression)
...
+        if ($result == null) {
+          ${ev.isNull} = true;
+        } else {
+          ${ev.value} = $result;
+        }
+      """
+    })
+  }
+
+  override def prettyName: String = "gcd"
```
Comment (bojana-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala]: This is redundant. Single word function names are correctly handled by the base class. (Same for `lcm`)

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathExpressionsSuite.scala [hunk 4]
```diff
@@ -488,6 +488,68 @@ class MathExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
     checkConsistencyBetweenInterpretedAndCodegen(Factorial.apply _, IntegerType)
   }
 
+  test("gcd") {
+    checkEvaluation(Gcd(Literal(24L), Literal(36L)), 12L)
+    checkEvaluation(Gcd(Literal(36L), Literal(24L)), 12L)
+    checkEvaluation(Gcd(Literal(17L), Literal(5L)), 1L)
+    // The result is never negative, whichever inputs are.
```
Comment (bojana-db) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathExpressionsSuite.scala]:
```suggestion
    // The result is non-negative regardless of the inputs' signs.
```
(Same for other occurrences.)

**Review by bojana-db — COMMENTED** [review]
LGTM, left a few comments

File: sql/core/src/test/resources/sql-tests/inputs/math.sql [hunk 5]
```diff
@@ -148,3 +148,37 @@ SELECT POSITIVE(25.5);
 SELECT POSITIVE("25.5");
 SELECT POSITIVE("invalid");
 SELECT POSITIVE(null);
+
+-- Gcd
+SELECT gcd(24, 36);
+SELECT gcd(17, 5);
+-- The result is never negative, whichever inputs are.
```
Comment (bojana-db) [inline sql/core/src/test/resources/sql-tests/inputs/math.sql]:
```suggestion
-- The result is non-negative regardless of the inputs' signs.
```

File: sql/core/src/test/resources/sql-tests/inputs/math.sql [hunk 6]
```diff
@@ -148,3 +148,37 @@ SELECT POSITIVE(25.5);
...
+SELECT gcd(0, 0);
+SELECT gcd(0, -7);
+SELECT gcd(7, 0);
+SELECT gcd(null, 5);
+SELECT gcd(5, null);
+
+-- Lcm
+SELECT lcm(4, 6);
+SELECT lcm(17, 5);
+-- The result is never negative, whichever inputs are.
```
Comment (bojana-db) [inline sql/core/src/test/resources/sql-tests/inputs/math.sql]:
```suggestion
-- The result is non-negative regardless of the inputs' signs.
```

**Review by bojana-db — APPROVED** [review]
LGTM, thank you for the feature!

Left some suggestions for grammatically awkward comments.

**Review by uros-b — APPROVED** [review]

**Comment by uros-b** [comment]
Thank you @Vivek1106-04 and @bojana-db!

**Comment by uros-b** [comment]
Let's loop in @Yicong-Huang and/or @HyukjinKwon for PySpark as well ^^

**Comment by HyukjinKwon** [comment]
cc @cloud-fan

**Comment by uros-b** [comment]
Also, please resolve conflicts @Vivek1106-04
